OpenMeteo/fetch_OpenMeteo_data.py

```diff
@@ -0,0 +1,120 @@
+import openmeteo_requests  # Importing required libraries
+import pandas as pd
+import requests_cache
+from retry_requests import retry
+
+
+class WeatherDataFetcher:
+    BASE_URL = "https://api.open-meteo.com/v1/"  # Base URL for OpenMeteo API
+
+    def __init__(self):
+        # Initialize the WeatherDataFetcher class
+        # Setup the Open-Meteo API client with cache and retry on error
+        cache_session = requests_cache.CachedSession(".cache", expire_after=3600)
+        retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
+        self.openmeteo = openmeteo_requests.Client(session=retry_session)
+
+    def fetch_forecast_data(self, NWP, params):
+        # Fetch weather data from OpenMeteo API for the specified model (NWP) and parameters
+        url = f"https://api.open-meteo.com/v1/{NWP}"  # Construct API URL
+        try:
+            responses = self.openmeteo.weather_api(url, params=params)  # Get weather data
+            return responses[0]  # Return the first response (assuming only one location)
+        except openmeteo_requests.OpenMeteoRequestsError as e:
+            # Handle OpenMeteoRequestsError exceptions
+            if "No data is available for this location" in str(e):
+                print(f"Error: No data available for the location for model '{NWP}'.")
+            else:
+                print(f"Error: {e}")
+            return None
+
+    def fetch_historical_data(self, params):
+        # Fetch historical weather data from OpenMeteo API
+        BASE_URL = "https://archive-api.open-meteo.com/v1/archive"
+        try:
+            responses = self.openmeteo.weather_api(BASE_URL, params=params)
+            return responses[0] if responses else None
+        except ValueError as e:
+            print(f"Error: {e}")
+            return None
+
+    def process_hourly_data(self, response):
+        # Process hourly data from OpenMeteo API response
+        # Extract hourly data from the response
+        hourly = response.Hourly()
+
+        # Extract variables
+        hourly_variables = {
```

Comment:
Ideally, the variables that are extracted are not hardcoded, but can be passed in as arguments.

Reply:
surely will do that, but could you please provide guidance on which variables should be included in the Xarray Dataset?

Reply:
Sorry, this reply slipped through, but I would go with by default, all available ones, and make one of the arguments a list of parameter names. I think there should be a way to get all the available parameters for a model from the API or something?